### PR TITLE
[DOC] Using External Libraries in Relay

### PR DESCRIPTION
**Pull request by SiNZeRo**
please help to review @tqchen

### CODE REVIEW

File: tutorials/relay/using_external_lib.py [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+
+This is a short tutorial on how to use external libraries such as cuDNN, or cuBLAS with Relay.
+
+Relay uses TVM internally to generate target specific code. For example, with cuda backend TVM generates cuda kernels for all layers in the user provided network.
+But sometimes it is also helpful to incorporate external libraries developed by various vendors into Relay.
+Luckily, TVM has a mechanism to transparently call into these libraries.
+For Relay users, all we need to do is just to set a target string appropriately.
+
+Before we can use external libraries from Relay, your TVM needs to be built with libraries you want to use.
+For example, to use cuDNN, USE_CUDNN option in tvm/make/config.mk needs to be enabled, and cuDNN include and library directories need to be specified.
```
Comment (masahi) [inline tutorials/relay/using_external_lib.py]: can you update this line for cmake? we don't use config.mk anymore.

Comment (SiNZeRo) [reply tutorials/relay/using_external_lib.py]: sure

File: tutorials/relay/using_external_lib.py [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+ctx = tvm.context(target, 0)
+data = np.random.uniform(-1, 1, size=data_shape).astype("float32")
+module = runtime.create(graph, lib, ctx)
+module.set_input(**params)
+module.set_input("data", data)
+module.run()
+out_shape = (batch_size, out_channels, 224, 224)
+out = module.get_output(0, tvm.nd.empty(out_shape))
+out_cudnn = out.asnumpy()
+exit(-1)
```
Comment (masahi) [inline tutorials/relay/using_external_lib.py]: what is this?

File: tutorials/relay/using_external_lib.py [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+######################################################################
+# Build and run with cuda backend
+# -------------------------------
+# We build and run this network with cuda backend, as usual.
+# By setting the logging level to DEBUG, the result of Relay graph compilation will be dumped as pseudo code.
+import logging
+logging.basicConfig(level=logging.DEBUG) # to dump TVM IR after fusion
+
+target = "cuda"
+with relay.build_config(opt_level=3):
```
Comment (masahi) [inline tutorials/relay/using_external_lib.py]: let's remove this line, to be consistent with the original doc. `opt_level=3` causes batch norm to be computed at compile time, it disappears from the psuedo code below.

File: tutorials/relay_quick_start.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
+"""
```
Comment (masahi) [inline tutorials/relay_quick_start.py]: what original tutorial correspond to this one?

Comment (SiNZeRo) [reply tutorials/relay_quick_start.py]: @masahi https://docs.tvm.ai/tutorials/nnvm_quick_start.html#sphx-glr-tutorials-nnvm-quick-start-py

File: tutorials/relay/using_external_lib.py (hunk 3 above)
Comment (SiNZeRo) [reply tutorials/relay/using_external_lib.py]: should I also remove the config of opt_level when building with `cuda -libs=cudnn`

Comment (masahi) [reply tutorials/relay/using_external_lib.py]: yes. Let's keep it simple

File: tutorials/relay_quick_start.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+.. _tutorial-relay-quick-start:
+Quick Start Tutorial for Compiling Deep Learning Models
+======================================================e
+**Author**: `Yao Wang <https://github.com/kevinthesun>`_, `Truman Tian <https://github.com/SiNZeRo>`_
+
+This example shows how to build a neural network with Relay python frontend and
+generate runtime library for Nvidia GPU with TVM.
```
Comment (masahi) [inline tutorials/relay_quick_start.py]: generates a runtime library

File: tutorials/relay_quick_start.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+batch_size = 1
+num_class = 1000
+image_shape = (3, 224, 224)
+data_shape = (batch_size,) + image_shape
+out_shape = (batch_size, num_class)
+
+net, params = relay.testing.resnet.get_workload(
+    num_layers=18, batch_size=batch_size, image_shape=image_shape)
+print(net.astext(show_meta_data=True))
```
Comment (masahi) [inline tutorials/relay_quick_start.py]: Shouldn't `show_meta_data` be False? I remember it dumps a lot of junk with `show_meta_data=True`

File: tutorials/relay_quick_start.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+# for this graph on the target hardware, and the parameter blobs of
+# the model. During the compilation, Relay does the graph-level
+# optimization while TVM does the tensor-level optimization, resulting
+# in an optimized runtime module for model serving.
+#
+# We'll first compile for Nvidia GPU. Behind the scene, `relay.build_module.build`
+# first does a number of graph-level optimizations, e.g. pruning, fusing, etc.,
+# then registers the operators (i.e. the nodes of the optimized graphs) to
+# TVM implementations to generate a `tvm.module`.
+# To generate the module library, TVM will first transfer the High level IR
```
Comment (masahi) [inline tutorials/relay_quick_start.py]: High -> high

File: tutorials/relay_quick_start.py (hunk 6 above)
Comment (SiNZeRo) [reply tutorials/relay_quick_start.py]: It's just a hint for using flag `show_meta_data` to show more information than simply `print (net)`

Let's set it to False to make the output clean :)

Comment (masahi) [reply tutorials/relay_quick_start.py]: ok, I tried locally and for this network the outputs of astext don't look much different when 
show_meta_data is True or False

**Comment by masahi** [comment]
Thanks I tested two tutorials locally. After minor typo fix and CI, I can merge this

**Comment by SiNZeRo** [comment]
> Thanks I tested two tutorials locally. After minor typo fix and CI, I can merge this

thanks for help

**Comment by masahi** [comment]
Thanks @SiNZeRo this is merged.

**Pull request merged**